src/main/java/de/tum/cit/aet/artemis/buildagent/service/SharedQueueProcessingService.java

```diff
@@ -350,6 +393,71 @@ private void processBuild(BuildJobQueueItem buildJob) {
         });
     }
 
+    private void pauseBuildAgent() {
+        if (this.isPaused) {
+            log.info("Build agent is already paused");
+            return;
+        }
+
+        log.info("Pausing build agent with address {}", hazelcastInstance.getCluster().getLocalMember().getAddress().toString());
+
+        this.isPaused = true;
+        this.removeListener();
+        if (this.scheduledFuture != null && !this.scheduledFuture.isCancelled()) {
+            this.scheduledFuture.cancel(false);
+        }
+        updateLocalBuildAgentInformation();
+
+        log.info("Gracefully cancelling running build jobs");
+
+        if (buildJobManagementService.getRunningBuildJobIds().isEmpty()) {
+            log.info("No running build jobs to cancel");
+        }
+        else {
+            // Sleep for 10 seconds to allow the build jobs to be finished. If they are not finished, they will be cancelled.
+            try {
+                Thread.sleep(pauseGracePeriodSeconds * 1000L);
```

Comment:
Can we also wait for the build job with a timeout of `pauseGracePeriodSeconds * 1000L`? This way we do not have to wait the full time in case the build job finishes earlier.